### PR TITLE
fix(loans): harden circulation edge cases from a full flow review + 100 guard tests

### PR DESCRIPTION
**Pull request by fabiodalez-dev**
## Context

I ran a full review of the loan/reservation system — state machine and entry points, overlap/capacity/copy allocation, automations and emails, availability and cache coherence — cross-checked against the behavioral evidence (all 22 existing circulation unit suites + the full 21-step E2E lifecycle, green before and after).

**The core is aligned**: both user entry points converge on the same `LoanRequestGate` under the same canonical lock; triggers come from a single source consumed by both installer and Updater; `@pinakes_application_date` is bound on every writer connection; `pickup_deadline` is never NULL (#366); `headQueuePos` participates in the capacity gate (#157); availability is never served from cache and every write path self-invalidates.

The review also surfaced a cluster of real edge-case bugs. This PR fixes all of them and adds **106 guard checks** plus an NCIP replay E2E regression.

## Fixes

**Silently lost emails for archived titles (#381 class)** — the sweeps deliberately expire circulation on soft-deleted books, but four notification fetches still filtered `deleted_at`: the loan expired, the copy was freed, and the borrower was never told. Fixed (with `CI-SOFT-DELETE-EXEMPT` annotations, matching the already-fixed manual sibling): `sendPickupExpiredNotification`, `sendReservationExpiredNotification`, `sendLoanReturnedNotification`, `notifyAdminsOverdue`, and the admin cancel-via-edit fetch in `ReservationsAdminController`.

**Copy-state drift on admin pickup cancel** — `cancelPickup` released any copy not in a broken state, including one in `prestato` physically out under another open loan (the reassignment service deliberately binds those to disjoint future windows). It now releases only a copy actually held for the pickup (`stato='prenotato'`), mirroring the user twin's guarded UPDATE.

**Copy-state drift on automated pickup expiry** — `checkExpiredPickups` had the same permissive release rule and could turn a copy already physically out under a disjoint open loan from `prestato` back into `disponibile`. The sweep now releases and reassigns only a copy it atomically transitions from `prenotato`.

**Allocator parity (#384)** — the reassignment allocator now shares the canonical preference ordering (prefer the copy with no later commitment), so repointing a hold never takes the copy a scheduled loan already needs. Its lenient date-overlap rule is kept and documented as a deliberate divergence: a reassignment may stack onto disjoint commitments because the alternative is cancelling an already-promised hold (`tests/loan-edge-cases.unit.php` encodes that contract — and caught my first, over-strict attempt).

**NCIP replay safety** — Pinakes currently exposes a title-level `libri.id` as NCIP `ItemId`, not a physical-copy identifier, and `CheckOutItem` has no request-idempotency key. Therefore NCIP checkouts intentionally keep the same-title policy strict even when the interactive staff UI enables multiple same-title loans: a partner retry returns `duplicate-request` instead of consuming a second copy. A real E2E runs with the multiple-loan setting enabled and proves that only one active loan is created.

**Consistency/robustness** — `checkExpiredPickups` clears `pickup_deadline` on the terminal row and passes the elapsed deadline to the email explicitly; rejected-loan fallback reasons render in the recipient's locale; `PickupDeadlineBackfill` binds `DateHelper::today()` instead of SQL `CURDATE()`; legacy `scripts/maintenance.php` no longer nests transactions (TXN-003) and flushes deferred promotion emails post-commit.

## Tests (106 checks, 4 new suites + NCIP E2E)

| Suite | Checks | Covers |
|---|---|---|
| `loan-archived-title-notifications` | 28 | Every fixed sender reaches archived-title borrowers (SMTP-free probes with negative controls); deadline plumbing; recipient-locale fallback |
| `loan-expiry-sweeps-behavior` | 31 | Pickup/reservation expiry transitions, guarded copy release (including a `prestato` negative control), availability recalc, idempotent re-runs, waitlist promotion, per-copy trigger semantics, multiplicity policy |
| `loan-allocator-384-parity` | 23 | Behavioral preference/hard-rule checks via reflection, headQueuePos capacity gate, allocator parity, and method-scoped NCIP strict-policy guards |
| `loan-backfill-and-invariant-guards` | 24 | Backfill behavior (cap/idempotency/app-TZ), legacy-script discipline, trigger single-source invariants, app-date binding, cancel-path release guards |

The NCIP browser suite adds a 32nd scenario that sends the same `CheckOutItem` twice with `allow_multiple_loans_same_book=1` and verifies `duplicate-request` plus exactly one active loan.

Stability: sandboxed fixtures with FK-safe cleanup, unique run markers, dates derived from the application today (no midnight-TZ flakes), no FULLTEXT/APCu/LiteSpeed dependencies. Each behavioral check is discriminating — it fails on the pre-fix code.

## Verification

- All 26 existing circulation suites green after the fixes
- New suites pass 3 consecutive runs
- NCIP replay E2E green against the local server
- `ci-quality-local.sh` green (PHPStan L5 full tree, soft-delete gate with the new exemptions, strict no-skip unit run)

## Summary by CodeRabbit

- **Nuove funzionalità**
  - Migliorata la riassegnazione delle copie, privilegiando quelle con impegni futuri più distanti.
  - Le notifiche vengono inviate anche per titoli archiviati.
  - I motivi di rifiuto sono localizzati nella lingua del destinatario.
  - Aggiornato il plugin NCIP Server alla versione 1.0.5.

- **Correzioni**
  - Evitata la liberazione errata di copie già prestate o disponibili.
  - Gestite correttamente scadenze, notifiche e date dei ritiri.
  - Migliorata la coerenza nelle promozioni dalla lista d’attesa.

- **Test**
  - Aggiunti test per allocazione copie, notifiche, scadenze, riassegnazioni e titoli archiviati.

### CODE REVIEW

File: app/Controllers/ReservationsAdminController.php [hunk 1]
```diff
@@ -327,10 +327,15 @@ public function update(Request $request, Response $response, mysqli $db, int $id
...
+                // CI-SOFT-DELETE-EXEMPT: the cancellation notice must reach the
+                // user even when the reserved title was archived meanwhile — the
+                // sibling cancel paths (cancelReservation, rejectLoan) are
+                // deliberately exempt for the same reason; filtering here made
+                // this one path silently skip the email.
                 $notificationStmt = $db->prepare(
                     "SELECT CONCAT(u.nome, ' ', u.cognome) AS utente_nome, u.email, l.titolo
                      FROM utenti u
-                     JOIN libri l ON l.id = ? AND l.deleted_at IS NULL
+                     JOIN libri l ON l.id = ?
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/ReservationsAdminController.php]: You're right — the update path already rejects archived titles with the book_not_found gate before this fetch can run, so the exemption was unreachable dead code. Reverted: the fetch keeps the standard deleted_at filter, and the test now guards the real invariant (up-front gate + consistent filter). Fixed in 14158623.

File: app/Support/NotificationService.php [hunk 2]
```diff
@@ -1485,10 +1488,13 @@ public function sendLoanRejectedNotification(int $loanId, string $reason = ''):
...
             $stmt->close();
 
+            // The fallback reason must render in the RECIPIENT's locale, not the
+            // acting admin's session locale (same pattern as the other senders).
+            $recipientLocale = $this->resolveRecipientLocale((string) $loan['utente_email']);
             $variables = [
                 'utente_nome' => $loan['utente_nome'],
                 'libro_titolo' => $loan['libro_titolo'],
-                'motivo_rifiuto' => $reason ?: __('Nessun motivo specificato')
+                'motivo_rifiuto' => $reason ?: $this->translateInLocale('Nessun motivo specificato', $recipientLocale)
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Support/NotificationService.php]: Fixed in 14158623 — both senders now use $reason !== '' so a literal "0" is preserved, same criterion as sendPickupCancelledNotification.

File: scripts/maintenance.php [hunk 3]
```diff
@@ -105,6 +105,11 @@
 \App\Support\DateHelper::synchronizeDatabaseSession($db);
 
 $reservationManager = new ReservationManager($db);
+// TXN-003: this script wraps each book in its own transaction below, so the
+// manager must NOT begin a nested one — mysqli begin_transaction() inside an
+// open transaction implicitly COMMITS the outer one, silently breaking the
+// FOR UPDATE serialization this loop relies on.
+$reservationManager->setExternalTransaction(true);
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply scripts/maintenance.php]: Fixed in 14158623 — external-transaction mode is switched on only AFTER cancelExpiredReservations() (which owns its transaction again), and its queued notifications are flushed immediately so a zero-book loop can no longer strand them.

File: tests/loan-archived-title-notifications.unit.php [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+$svc = new ArchivedProbeNotificationService($db);
+$svc->sendLoanReturnedNotification($loanId);
+$check($svc->localeResolved === true, '08 return-confirmation email fetches the loan for a soft-deleted book');
+$check($svc->capturedEmail === $email, '09 return-confirmation resolves the borrower email for the archived title');
+
+// ── 10. Admin overdue alert also reaches past the JOIN for archived titles ───
+$svc = new ArchivedProbeNotificationService($db);
+$svc->notifyAdminsOverdue($loanId);
+$check($svc->localeResolved === true, '10 admin overdue alert fetches the loan for a soft-deleted book');
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/loan-archived-title-notifications.unit.php:210]: Confirmed — this was exactly the CI failure on the DB-matrix jobs (fresh schema, no admin users). The suite now seeds a sandbox admin so sendToAdmins reaches its recipient loop deterministically in every environment. Fixed in 14158623.

File: tests/loan-allocator-384-parity.unit.php [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+$check($gateAllocPos !== false
+    && str_contains($gateAllocSrc, 'p.data_prestito <= ?')
+    && !str_contains($gateAllocSrc, 'data_scadenza >= ?'),
+    '19 the gate keeps the HARD rule: any commitment starting before the window end rejects');
+
+// NCIP desk operations share the same policy and preference (no drifting clones).
+$ncipSrc = (string) file_get_contents($root . '/storage/plugins/ncip-server/NcipServerPlugin.php');
+$check(str_contains($ncipSrc, "'9999-12-31'"),
+    '20 the NCIP checkout allocator carries the #384 preference sentinel');
+$check(substr_count($ncipSrc, 'LoanMultiplicityPolicy') >= 2,
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/loan-allocator-384-parity.unit.php]: Fixed in a261ce09 — checks 20-23 now slice each method body (anchor at the definition, bounded at the next method, failing if the anchor disappears) and assert the copy-binding call in createLoanAtomic and the strict call in createRequestItemNcip specifically. Proven non-vacuous by mutation: removing either call makes check 21 fail.
